deps/sharedlibpp/src/SharedLibraryFactory.cpp

```diff
@@ -64,6 +84,38 @@ shlibpp::SharedLibraryFactory::Private::Private(int32_t startCheck,
     memset(&api, 0, sizeof(SharedLibraryClassApi));
 }
 
+std::vector<std::string> shlibpp::SharedLibraryFactory::Private::platformSpecificLibName(const std::string& library)
+{
+
+#if defined(_WIN32)
+    return {library + ".dll", library + "d.dll", "lib" + library + ".dll"};
```

Comment:
I am a bit confused by this. Why is `"lib" + library + ".dll"` is included and `"lib" + library + "d.dll"` not? 

Furthermore, if the order of the vector indicates the order in which the candidates are searched for, I think that depending on whether `BlockFactory` is compiled in release or in debug, the order needs to be changed to make sure that in Debug builds the `library + "d.dll"` is searched first. It could make sense to still have the normal one as a fallback, but if both are found and BlockFactory is compiled in debug, for sure `library + "d.dll"` is the one to load.

Reply:
> I am a bit confused by this. Why is `"lib" + library + ".dll"` is included and `"lib" + library + "d.dll"` not?

Because this case happens only if you don't use the VS generator. This is for instance the case with Ninja, and since it is not a multi-config generator the debug suffix is not used.

For what concern the order, yes, on windows it actually makes sense to define it wrt to the `CMAKE_BUILD_TYPE`. There's always the caveat of linking Release code to Release code, and I think this applies also to plugin libraries.

Reply:
> Because this case happens only if you don't use the VS generator. This is for instance the case with Ninja, and since it is not a multi-config generator the debug suffix is not used.

Cool. So if you use Ninja and compile BlockFactory in Debug, the libraries are compiled with the `lib` prefix and without the `d` suffix, even if `CMAKE_DEBUG_POSTFIX` is set to `d` in https://github.com/robotology/blockfactory/blob/5a0b977cf0c6fe7450ee73a57674d59174ae891d/CMakeLists.txt#L95 ? 

Ack then, even if it really seems to be a CMake+Ninja+MSVC bug.

Reply:
> For what concern the order, yes, on windows it actually makes sense to define it wrt to the `CMAKE_BUILD_TYPE`. There's always the caveat of linking Release code to Release code, and I think this applies also to plugin libraries.

Nitpick: `CMAKE_BUILD_TYPE` is ignored for  multiple config generator. Do you have any idea of where this logic in  YARP? As far as I know, in YARP that logic is is working fine.

Reply:
> Ack then, even if it really seems to be a CMake+Ninja+MSVC bug.

In my tests I was using g++ with Ninja (check [travis](https://travis-ci.com/diegoferigo/blockfactory/jobs/173692501#L55)). I am not 100% sure of the outcome of the combo of MSVC compiler + Ninja generator. I guess it generates the same things of using just MSVC.

Reply:
Ahh, sorry I did not understand we wanted to support Mingw. 
In that case, in that case it make sense, the `CMAKE_DEBUG_POSTFIX` is only set for `MSVC`, not for `WIN32`.

Reply:
> Nitpick: CMAKE_BUILD_TYPE is ignored for multiple config generator. Do you have any idea of where this logic in YARP? As far as I know, in YARP that logic is is working fine.

You mean in `shlibpp`? I didn't find it, I think that this portion of code already handles it:

https://github.com/robotology/blockfactory/blob/7b9033c133488b648fc04c509c24c764a7c3cd64/deps/sharedlibpp/src/SharedLibrary.cpp#L56-L69

In the example of `shlibpp` you must pass the OS-specific name (see [test](https://github.com/robotology-playground/sharedlibpp/blob/0a461dd78b7d84b745e33a5f6ade54873047824b/examples/math_test.cpp#L19-L23)). I added the support of doing it automatically in this PR.

Actually, I just realized that if the library is not in the extended search path, [I pass only the libname](https://github.com/robotology/blockfactory/blob/74a7a0e6db7e05c5ae7c3d82f7c8f06bdc17081c/deps/sharedlibpp/src/SharedLibraryFactory.cpp#L132-L134). This case would cover the case where the library is in the linker search path, and I'm not sure `dlopen` can be called without specifying the prefix (lib) and suffix (.so) of the library.

Reply:
> No, I meant in YARP. I found it, it is in https://github.com/robotology/yarp/blob/master/src/libYARP_OS/src/YarpPlugin.cpp#L69 .

:+1: 

Btw I think that YARP code does not work well with `icc` ^^ I'm not sure tho if we support it.

> Regarding mingw, in theory the mingw would be searched first if BlockFactory is compiled with mingw, as Mingw's and MSVC's C++ ABI are not compatible. However, the trick of <name>d.dll, name.dll to distinguish the various ABI is just a dirty trick (there are a few other differences that should be considered to fully determine the ABI compatibility), so I guess it is not tragic if we search it later, as in most cases Mingw-libraries and MSVC-libraries are not installed in the same prefix.

These are edge cases at this stage. We never stated the we support Mingw and the installation instructions on Windows only mention MSVC. I think we can postpone safely these cases.